src/time/Waveform.hpp

```diff
@@ -0,0 +1,122 @@
+  void storeAtFirstSample(const Eigen::VectorXd &values);
+
+  /**
+   * @brief Updates all entries in _timeWindowsStorage with given values. Often used to initialize a new Waveform with given data.
+   * @param values Sample used for all time window
+   */
+  void storeAtAllSamples(const Eigen::VectorXd values);
+
+  /**
+   * @brief Shifts all entries in _timeWindows. The new entry is initialized as the value from the last window (= constant extrapolation). Usually called, when moving to the next time window.
+   */
+  void moveToNextWindow();
+
+  /**
+   * @brief Evaluate Waveform at specific point in time. Uses interpolation with Waveform's interpolation order, if necessary.
+   * @param normalizedDt Time where the sampling inside the window happens. Only allows values between 0 and 1. 0 refers to the beginning of the window and 1 to the end.
+   * @return Value of Waveform at time normalizedDt.
+   */
+  Eigen::VectorXd sample(const double normalizedDt);
+
+private:
+  /// Set by initialize. Used for consistency checks.
+  bool _storageIsInitialized = false;
+
+  /// Stores values for several time windows.
+  Eigen::MatrixXd _timeWindowsStorage;
+
+  /// interpolation order for this waveform
+  const int _interpolationOrder;
+
+  /// number of stored samples in _timeWindowsStorage
+  int _numberOfStoredSamples;
+
+  mutable logging::Logger _log{"time::Waveform"};
+
+  /**
+   * @brief Get number of values per sample in time stored by this waveform.
+   * @return Number of values per sample.
+   */
+  int valuesSize();
+
+  /**
+   * @brief Get maximum number of samples in time this waveform can store.
+   * @return Maximum number of samples.
+   */
+  int maxNumberOfStoredSamples();
+
+  /**
+   * @brief Updates entry in _timeWindowsStorage corresponding to a given sampleIndex with given values.
+   * @param values Input sample.
+   * @param sampleIndex Index of sample to be updated to given input sample.
+   */
+  void storeAt(const Eigen::VectorXd values, int sampleIndex);
+
+  /**
+   * @brief Computes which order may be used for interpolation.
+   * 
+   * Order of interpolation is determined by number of stored samples and maximum order defined by the user.
+   * Example: If only two samples are available, the maximum order we may use is 1, even if the user demands order 2.
+   *
+   * @param requestedOrder Order requested by the user.
+   * @param numberOfAvailableSamples Samples available for interpolation.
+   * @return Order that may be used.
+   */
+  int computeUsedOrder(int requestedOrder, int numberOfAvailableSamples);
+
+  /**
+   * @brief Interpolates values inside current time window using _timeWindowsStorage and an interpolation scheme of the order of this Waveform.
+   * @param normalizedDt time where the sampling inside the window happens. 0 refers to the beginning of the window and 1 to the end.
+   * @return Interpolated value at time normalizedDt.
+   */
+  Eigen::VectorXd interpolate(const double normalizedDt);
```

Comment:
Is this additional function really necessary?

Reply:
Let's remove it for now. Later it might be good to move it outside of `Waveform::sample` again, but currently it is simple enough to be inlined (and this saves us a bit of documentation and maintainment). c47c5afa60112ef25bf4b0eeb053afb2bc10d37b